### ISSUE TITLE
[BUG] pathlib.Path is not serializable

### ISSUE THREAD

**Issue by devtud** [tiangolo/fastapi#563]
**Describe the bug**
If the api response contains fields of type `pathlib.Path`, an error is raised:
 
```bash
ValueError: [KeyError(<class 'pathlib.PosixPath'>,), TypeError("'PosixPath' object is not iterable",), TypeError('vars() argument must have __dict__ attribute',)]
```

**To Reproduce**
I wrote a test for this bug which fails with the above ^ error.

```python
#tests/test_serialize_path.py

from pathlib import Path

from pydantic import BaseModel
from starlette.testclient import TestClient

from fastapi import FastAPI


class ContainsPathProperty(BaseModel):
    path: Path


app = FastAPI()


@app.get("/", response_model=ContainsPathProperty)
async def get_main():
    obj = ContainsPathProperty(path=Path('/some/random/path'))
    return obj


client = TestClient(app)


def test_serialize_path():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {'path': '/some/random/path'}
```

**Expected behavior**
The test should pass, which means a 200 response is returned and the response is:

```json
{
    "path": "/some/random/path"
}
```

**Environment:**
 - OS: Linux / Windows
 - FastAPI Version 0.38.1

- Python version 3.6.9

**Additional context**

**Solution 1**

A possible fix would be importing and using the encoder from `pydantic` which knows to encode `Path` objects:

```python
# pydantic/json.py

def pydantic_encoder(obj: Any) -> Any:
    ......
    elif isinstance(obj, Path):
        return str(obj)
    ......
```

The Fastapi's `jsonable_encoder(....)` would be changed from this:

```python
# fastapi/encoder.py
....
from pydantic.json import ENCODERS_BY_TYPE
....

def jsonable_encoder(....) -> Any:
    .......
    try:
        if custom_encoder and type(obj) in custom_encoder:
            encoder = custom_encoder[type(obj)]
        else:
            encoder = ENCODERS_BY_TYPE[type(obj)]
        return encoder(obj)
    except KeyError as e:
        .....
    ......
```

to this:

```python
# fastapi/encoder.py
....
from pydantic.json import pydantic_encoder
....

def jsonable_encoder(....) -> Any:
    .......
    try:
        if custom_encoder and type(obj) in custom_encoder:
            encoder = custom_encoder[type(obj)]
        else:
            encoder = pydantic_encoder
        return encoder(obj)
    except KeyError as e:
        .....
    ......
```

The test passes but `pydantic_encoder()` has a downside: the string obtained by `str(Path(...))` is not the same on Windows (eg: `\some\random\path`) vs Linux (`/some/random/path`).

** Solution 2 **

As I think it is very important that a Fastapi app return the same responses no matter the platform on which it's deployed, another solution is handling the `pathlib.Path` encoding directly in Fastapi's `jsonable_encoder()`, using `Path(...).as_posix()` instead of `str(Path(...))` like:

```python
# fastapi/encoders.py

def jsonable_encoder(.....) -> Any:
    ......
    if isinstance(obj, Path):
        return obj.as_posix()
    .......
```

What do you guys think?

**Comment by euri10** [comment]
imho it's an issue upstream, if you switch to `from pydantic import Path` you get 
```
                except Exception as e:
                    errors.append(e)
>                   raise ValueError(errors)
E                   ValueError: [KeyError(<class 'pathlib.PosixPath'>), TypeError("'PosixPath' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]
```

which suggests a custom encoder wasnt written for `PosixPath` (see ENCODERS_BY_TYPE)

One workaround is this:
```
from pathlib import PosixPath

from pydantic import BaseModel, Path
from starlette.testclient import TestClient

from fastapi import FastAPI


class ContainsPathProperty(BaseModel):
    path: Path

    class Config:
        json_encoders = {
            PosixPath: lambda v: str(v)
        }


app = FastAPI()


@app.get("/", response_model=ContainsPathProperty)
async def get_main():
    obj = ContainsPathProperty(path=Path('/some/random/path'))
    return obj


client = TestClient(app)


def test_serialize_path():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {'path': '/some/random/path'}
```

**Comment by devtud** [comment]
Thanks for the quick response @euri10 and for the workaround.

> which suggests a custom encoder wasnt written for PosixPath (see ENCODERS_BY_TYPE)

Please see [pydantic_encoder()](https://github.com/samuelcolvin/pydantic/blob/9ffa311f8fd54ad2a5619477a0612f99c7bc7ae4/pydantic/json.py#L44) which is the encoder used by pydantic. Fastapi uses a custom one based on the mapping mentioned by you (in which, indeed, PosixPath is not handled). But I don't consider this an upstream issue because Pydantic itself works ok. The only issue with it if it had been used in Fastapi would be that it encodes differently in Windows than Linux.

**Comment by euri10** [comment]
Iirc it's incorrect, effectively FastAPI uses its own encoder but it relies
on pydantic one, especially for other types, you'll see pretty much the
same loop on custom encoder.

Le mar. 24 sept. 2019 à 10:01 AM, devtud <notifications@github.com> a
écrit :

**Comment by euri10** [comment]
On top of that the windows /Linux discrepancy should be eliminated by
as_posix in the encoder I suggested, shouldn't it?

**Comment by devtud** [comment]
@euri10 , my reply was before seeing you edit. However, the workaround which also would solve the discrepancy would be:

```python
class ContainsPathProperty(BaseModel):
    path: Path

    class Config:
        json_encoders = {
            PosixPath: lambda v: v.as_posix(), // or lambda v: str(v)
            WindowsPath: lambda v: v.as_posix()
        }
```
Do you agree?

The first encoder will be executed on Linux/MacOS, and the second one on Windows.

However, this should be fixed in Fastapi (or Pydantic if considered).

**Comment by euri10** [comment]
I dont have windows sorry so cant really comment on that part, but your approach seems perfectly valid
I disagree this should be fixed in FastAPI and thinking about it in Pydatic too, the way you want to encode such an object may be different from one person to another, maybe pydantic could provide a default encoding, not even sure about that, @dmontagu will know.

**Comment by dmontagu** [comment]
It looks like you've got a workaround, but if you'd like standardized encoding of `Path`, I'd encourage you to create an issue for discussion in the pydantic repo. 

It seems like a reasonable request to me, though I could imagine some controversy over whether it should default to a posix format or `str`-casting -- given that, it might be best left to users to decide.

**Comment by GlennS** [comment]
Is this Pydantic fix relevant?

https://github.com/samuelcolvin/pydantic/issues/473
https://github.com/samuelcolvin/pydantic/pull/485

**Comment by devtud** [comment]
@GlennS , that fix is in `pydantic_encoder()` which is not used by Fastapi.

**Comment by tiangolo** [comment]
Thanks for the report @devtud .

Yes, that's a bug/feature request. We should have a case for it in `jsonable_encoder`.

Right after:

```Python
if isinstance(obj, Enum):
        return obj.value
```

**Comment by tiangolo** [comment]
This was fixed by @patrickmckenna in #978 :rocket: :bug: 

It's available in version `0.49.0` :tada:
